syntax = "proto3";

option java_package = "gr.grpc.generated";

service Explanations {
    rpc GetExplanation (stream ExplanationsRequest) returns (ExplanationsResponse);
    rpc Initialization (InitializationRequest) returns (InitializationResponse);
}


message InitializationRequest {
    string model_name = 1;
}

message TableModel {
    ExplanationsResponse content = 1;
}

message Model {
    repeated string feature_names = 1;
    map <string,ExplanationsResponse> plots = 2;
    map <string,TableModel> tables = 3;
}

message InitializationResponse {
    Model model = 1;
    Pipeline pipeline = 2;
}

message Pipeline {
    repeated string hyperparameter_names = 1;
    map <string,ExplanationsResponse> plots = 2;
    map <string,TableModel> tables = 3;
}

// ----------------- Explanations Fields ----------------------------


message ExplanationsRequest {

    string explanation_type = 1;
    string explanation_method = 2;

//------------------ Pipeline Explainability --------------------------------------------------

    // ----- PDP - ALE Fields -----

    string model = 3;
    string feature1 = 4;
    string feature2 = 5;

    // ----- Influence Functions Fields -----

    int32 num_influential = 6;

    // ----- Counterfactual Explanation Fields -----
    bytes proxy_dataset = 7;
    bytes query = 8;

//------------------ Model Explainability --------------------------------------------------

    // ------------ PDP Fields -------------------------
    // the other same as Pipeline Explainability plus
    string features = 9;

    // ------------ CounterfactualExplanations Fields -------------------------
    // same as Model explainability plus target target label 
    string target = 10;



    // Include other fields if needed based on the chosen explanation type
}


 message Features {
    string feature1 = 1;
    string feature2 = 2;
}

message Axis {
    string axis_name = 1;
    repeated string axis_values= 2;
    string axis_type = 3;
}

message ExplanationsResponse {
    string explainability_type = 1;
    string explanation_method = 2;
    string explainability_model = 3;
    string plot_name = 4;
    string plot_descr = 5;
    string plot_type = 6;
    Features features = 7;
    Axis xAxis = 8;
    Axis yAxis = 9;
    Axis zAxis = 10; 
    //map<string, repeated string> table_contents= 11;
}
